kratos/factories/base_factory.h

```diff
@@ -0,0 +1,270 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Vicente Mataix Ferrandiz
+//
+
+#if !defined(KRATOS_BASE_FACTORY_H_INCLUDED )
+#define  KRATOS_BASE_FACTORY_H_INCLUDED
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "includes/kratos_parameters.h"
+#include "includes/kratos_components.h"
+#include "includes/shared_pointers.h"
+#include "solving_strategies/strategies/explicit_solving_strategy.h"
+#include "solving_strategies/builder_and_solvers/explicit_builder.h"
+#include "spaces/ublas_space.h"
+
+namespace Kratos
+{
+///@name Type Definitions
+///@{
+
+typedef TUblasSparseSpace<double> SparseSpaceType;
+typedef TUblasDenseSpace<double> LocalSpaceType;
+
+///@}
+///@name Kratos Classes
+///@{
+
+/**
+ * @class BaseFactoryMethods
+ * @ingroup KratosCore
+ * @brief Here we define some common methods
+ * @details Defines the base class factory methods
+ * @author Vicente Mataix Ferrandiz
+ */
+class BaseFactoryMethods
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of BaseFactoryMethods
+    KRATOS_CLASS_POINTER_DEFINITION(BaseFactoryMethods);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /** Constructor.
+     */
+    explicit BaseFactoryMethods(){}
+
+    /** Destructor.
+     */
+    virtual ~BaseFactoryMethods(){}
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief This method checks if the linear class is registered
+     * @param rClassName The nanme of the class
+     * @return True if registered, false otherwise
+     */
+    virtual bool Has(const std::string& rClassName) const
+    {
+        KRATOS_ERROR << "Methods must be implemented in the base class" << std::endl;
+        return false;
+    }
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    /// Turn back information as a string.
+    virtual std::string Info() const
+    {
+        return "BaseFactoryMethods";
+    }
+
+    /// Print information about this object.
+    virtual void PrintInfo(std::ostream& rOStream) const
+    {
+        rOStream << this->Info() << std::endl;
+    }
+
+    /// Print object's data.
+    virtual void PrintData(std::ostream& rOStream) const
+    {
+//         rOStream << this->Info() << std::endl;
+    }
+};
+
+/**
+ * @class BaseFactory
+ * @ingroup KratosCore
+ * @brief Here we add the functions needed for the registration of class
+ * @details Defines the base class factory
+ * @author Vicente Mataix Ferrandiz
+ * @tparam TClass The class to create the factory
+ * @tparam TAuxiliarClass The auxiliar class to create the factory
+ */
+template<typename TClass, typename TAuxiliarClass = TClass>
+class BaseFactory
```

Comment:
hm I am thinking that the other classes (e.g. `BuilderAndSolver) don't start with `Base` so probably we should remove it here to stay consistent

Reply:
Base of base class

Reply:
=> `Factory`